versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,376 @@
+      this.value = value;
+    }
+
+    @Override
+    public <TDocument> BsonDocument toBsonDocument(Class<TDocument> clazz, CodecRegistry codecRegistry) {
+      // Intentionally don't use Updates.setOnInsert as that will result in issues encoding the value entity,
+      // due to codec lookups the MongoDB driver will actually encode the fields of the basic object, not the entity.
+      final BsonDocumentWriter writer = new BsonDocumentWriter(new BsonDocument());
+      writer.writeStartDocument();
+      writer.writeName("$setOnInsert");
+      // This intentionally doesn't use generics to avoid casting issues around captures.
+      Codec codec = codecRegistry.get(clazz);
+      codec.encode(writer, value, EncoderContext.builder().build());
+      writer.writeEndDocument();
+      return writer.getDocument();
+    }
+  }
+
+  private static final InsertManyOptions INSERT_UNORDERED = new InsertManyOptions().ordered(false);
+  private static final UpdateOptions UPDATE_UPSERT = new UpdateOptions().upsert(true);
+  private static final ReplaceOptions REPLACE_UPSERT = new ReplaceOptions().upsert(true);
+
+  // This should be retrievable via the API as well.
+  private final int paginationSize = 100000;
+  private final MongoStoreConfig config;
+  private final MongoClientSettings mongoClientSettings;
+
+  private MongoClient mongoClient;
+  private MongoDatabase mongoDatabase;
+  private final Map<ValueType, MongoCollection<? extends HasId>> collections;
+
+  /**
+   * Creates a store ready for connection to a MongoDB instance.
+   * @param config the configuration for the store.
+   */
+  public MongoDbStore(MongoStoreConfig config) {
+    this.config = config;
+    this.collections = new HashMap<>();
+    final CodecRegistry codecRegistry = CodecRegistries.fromProviders(
+        new CodecProvider(),
+        PojoCodecProvider.builder().automatic(true).build(),
+        MongoClientSettings.getDefaultCodecRegistry());
+    this.mongoClientSettings = MongoClientSettings.builder()
+      .applyConnectionString(new ConnectionString(config.getConnectionString()))
+      .codecRegistry(codecRegistry)
+      .writeConcern(WriteConcern.MAJORITY)
+      .build();
+  }
+
+  /**
+   * Gets a handle to an existing database or get a handle to a MongoDatabase instance if it does not exist. The new
+   * database will be lazily created.
+   * Since MongoDB creates databases and collections if they do not exist, there is no need to validate the presence of
+   * either before they are used. This creates or retrieves collections that map 1:1 to the enumerates in
+   * {@link com.dremio.nessie.versioned.store.ValueType}
+   */
+  @Override
+  public void start() {
+    mongoClient = MongoClients.create(mongoClientSettings);
+    mongoDatabase = mongoClient.getDatabase(config.getDatabaseName());
+
+    // Initialise collections for each ValueType.
+    collections.put(ValueType.L1, mongoDatabase.getCollection(config.getL1TableName(), L1.class));
```

Comment:
Can we make this cleaner? Maybe declare a map that is valuetype -> table name retrieval function then just iterate over it.